src/librustc/ty/constness.rs

```diff
@@ -67,15 +68,55 @@ impl<'tcx> TyCtxt<'tcx> {
 }
 
 
-pub fn provide(providers: &mut Providers<'_>) {
-    /// only checks whether the function has a `const` modifier
-    fn is_const_fn_raw(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
+pub fn provide<'tcx>(providers: &mut Providers<'tcx>) {
+    fn is_const_evaluatable(tcx: TyCtxt<'tcx, 'tcx>, def_id: DefId) -> bool {
+        // Const evaluability whitelist is here to check evaluability at the
+        // top level beforehand.
+        match tcx.fn_sig(def_id).abi() {
+            Abi::RustIntrinsic |
+            Abi::PlatformIntrinsic => {
+                match &tcx.item_name(def_id).as_str()[..] {
+                    | "size_of"
+                    | "min_align_of"
+                    | "needs_drop"
+                    | "type_id"
+                    | "bswap"
+                    | "bitreverse"
+                    | "ctpop"
+                    | "cttz"
+                    | "cttz_nonzero"
+                    | "ctlz"
+                    | "ctlz_nonzero"
+                    | "overflowing_add"
+                    | "overflowing_sub"
+                    | "overflowing_mul"
+                    | "unchecked_shl"
+                    | "unchecked_shr"
+                    | "rotate_left"
+                    | "rotate_right"
+                    | "add_with_overflow"
+                    | "sub_with_overflow"
+                    | "mul_with_overflow"
+                    | "saturating_add"
+                    | "saturating_sub"
+                    | "transmute"
+                    => true,
+
+                    _ => false
+                }
+            }
+            _ => false
+        }
+    }
+
+    /// Checks whether the function has a `const` modifier and intrinsics can be promotable in it
+    fn is_const_fn_raw<'tcx>(tcx: TyCtxt<'tcx, 'tcx>, def_id: DefId) -> bool {
         let hir_id = tcx.hir().as_local_hir_id(def_id)
                               .expect("Non-local call to local provider is_const_fn");
 
         let node = tcx.hir().get(hir_id);
         if let Some(fn_like) = FnLikeNode::from_node(node) {
-            fn_like.constness() == hir::Constness::Const
+            (fn_like.constness() == hir::Constness::Const) || is_const_evaluatable(tcx, def_id)
```

Comment:
rename `is_const_evaluatable` to `is_const_intrinsic` and check it first thing in `is_const_fn_raw`. If it's `true`, return `true`. (as suggested above in https://github.com/rust-lang/rust/pull/61835/files#r294078191)